### ISSUE TITLE
Data race in client.go

### ISSUE THREAD

**Issue by djoyner** [jcelliott/turnpike#118]
There appears to be a data race in client.go:

A write to the events map in `Subscribe`
(https://github.com/jcelliott/turnpike/blob/v2/client.go#L348):

```
    c.events[subscribed.Subscription] = &eventDesc{topic, fn}
```

Races with a read from the events map in `Receive`
(https://github.com/jcelliott/turnpike/blob/v2/client.go#L215):

```
        if event, ok := c.events[msg.Subscription]; ok {
```

**Comment by djoyner** [comment]
The above issue was flagged by the race detector running in some downstream code.  I already have a fix for the issue -- using a `sync.Mutex` to protect all three of the client's maps.  While preparing a PR I ran the existing unit tests and saw that the new [Golang 1.6 concurrent map write detector](https://golang.org/doc/go1.6#runtime) almost always panics.  I checked with `go test -race *.go` and there are many more races identified in other parts of the code.

I'd be happy to open a PR with my `client.go` fixes but I'm wondering if there is any plan to address the bigger set of concurrency issues in this package?  If that's already in progress I will hold off.

**Comment by beatgammit** [comment]
> I'd be happy to open a PR with my client.go fixes but I'm wondering if there is any plan to address the bigger set of concurrency issues in this package? If that's already in progress I will hold off.

I've been working on some fixes in my fork, one of which being this exact one you mention. There may need to be some breaking changes to properly fix the issues though. I'll probably merge in my changes soon (hopefully this week sometime) and then I'll turn my attention to the race detector.

I've been running with Go 1.4 for a few projects so I haven't been running the race detector. I recently started upgrading and fixing race issues. Unfortunately, I wasn't aware that in Go 1.4 everything was locked to a single thread, so it wasn't until switching to 1.6 that I started realizing how many data races there are. I've been going through and auditing my code looking for false assumptions (e.g. I thought maps were thread safe, but I was wrong).

So in short, yes I'm working on some fixes and I'm going to be auditing this project to make sure it's production ready.

**Comment by mattbaird** [comment]
I have a few concurrency fixes I would contribute, but it's on v1. I only recently started experimenting with v2.

**Comment by djoyner** [comment]
@beatgammit Good to know, thanks. @DaniloMoura1 and I were preparing to wade in to do exactly what you just described but since you're already working on it, we will wait for the update.
